ksqldb-rest-app/src/main/java/io/confluent/ksql/api/impl/BlockingPrintPublisher.java

```diff
@@ -0,0 +1,206 @@
+import io.vertx.core.WorkerExecutor;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ExecutionException;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.clients.consumer.ConsumerRecords;
+import org.apache.kafka.clients.consumer.KafkaConsumer;
+import org.apache.kafka.common.record.RecordBatch;
+import org.apache.kafka.common.utils.Bytes;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A query publisher that uses an internal blocking queue to store rows for delivery. It's currently
+ * necessary to use a blocking queue as Kafka Streams delivers message in a synchronous fashion with
+ * no back pressure. If the queue was not blocking then if the subscriber was slow the messages
+ * could build up on the queue eventually resulting in out of memory. The only mechanism we have to
+ * slow streams down is to block the thread. Kafka Streams uses dedicated streams per topology so
+ * this won't prevent the thread from doing useful work elsewhere but it does mean we can't have too
+ * many push queries in the server at any one time as we can end up with a lot of threads.
+ */
+public class BlockingPrintPublisher extends BasePublisher<String> {
+
+  private static final Logger log = LoggerFactory.getLogger(BlockingPrintPublisher.class);
+
+  public static final int SEND_MAX_BATCH_SIZE = 200;
+
+  private final WorkerExecutor workerExecutor;
+  private final KafkaConsumer<Bytes, Bytes> topicConsumer;
+  private final PrintTopic printTopic;
+  private final int limit;
+  private final long interval;
+  private final RecordFormatter formatter;
+  private final Duration disconnectCheckInterval;
+  private volatile boolean closed;
+  private volatile boolean started;
+
+  public BlockingPrintPublisher(final Context ctx,
+      final WorkerExecutor workerExecutor,
+      final ServiceContext serviceContext,
+      final Map<String, Object> consumerProperties,
+      final PrintTopic printTopic,
+      final Duration disconnectCheckInterval) {
+    super(ctx);
+    this.workerExecutor = Objects.requireNonNull(workerExecutor);
+    this.printTopic = printTopic;
+    this.limit = printTopic.getLimit().orElse(Integer.MAX_VALUE);
+    this.interval = printTopic.getIntervalValue();
+    this.formatter = new RecordFormatter(
+        serviceContext.getSchemaRegistryClient(),
+        printTopic.getTopic()
+    );
+    this.disconnectCheckInterval = disconnectCheckInterval;
+    this.topicConsumer = createTopicConsumer(serviceContext, consumerProperties, printTopic);
+  }
+
+  public Future<Void> close() {
+    if (closed) {
+      return Future.succeededFuture();
+    }
+    closed = true;
+    // Run async as it can block
+    ctx.runOnContext(v -> sendComplete());
+    executeOnWorker(topicConsumer::close);
```

Comment:
Does this actually need to be async? I guess it doesn't hurt since the default timeout is large; just wanted to understand the thinking here.

Reply:
Hmm, I don't think it has to be run async. I just followed the pattern for closing the publisher we had over in the QueryStreamHandler and thought it made sense because of the 30s timeout for the consumer.close() function.